README.pod
=pod

=head1 NAME

Attribute::Contract - Design by contract via Perl attributes

=head1 SYNOPSIS

    package Interface;
    use Attribute::Contract -types => [qw/Str slurpy ArrayRef/];

    sub do_smth :ContractRequires(Str, slurpy ArrayRef[Str]) :ContractEnsures(Str) {
        my $self = shift;
        my ($input_string, $array_ref_of_strings) = @_;

        return '...';
    }

    package Implementation;
    use base 'Interface';
    use Attribute::Contract;

    sub do_smth {
        my $self = shift;
        my ($input_string, $array_ref_of_strings) = @_;

        return 'ok';
    }

    Implementaion->do_smth('hi', 'there'); # works

    Implementaion->do_smth();              # croaks!
    Implementaion->do_smth(sub {});        # croaks!

=head1 DESCRIPTION

L<Attribute::Contract> by using Perl attributes allows you to specify contract
(L<Design by Contract|http://en.wikipedia.org/wiki/Design_by_contract>) for
every method in your class. You can check incoming and outgoing values by
specifying C<ContractRequires> and C<ContractEnsures> attributes.

It's the most useful for interfaces or abstract classes when you want to control
whether your implementation follows the same interface and respects the Liskov
substitution principle.

This module uses L<Type::Tiny> underneath so all the checks is done via that
module. Check it out for more documention on type validation.

Why attributes? They feel and look natural and are applied during compile time.

=head2 IMPORTING

When using L<Attribute::Contract> one may want to import various types in order
to check them in attributes. Types themselves are not imported into the current
module but rather used when compiling attributes.

=head3 Types

    package MyClass;
    use Attribute::Contract -types => [qw/ClassName Str/];

    sub static_method : ContractRequires(ClassName, Str) {
    };

    ...

    MyClass->static_method('string');

=head3 Type libraries

When types are complex or the description is too long attributes might get not
very readable. In this case one can use type libraries (implemented again via
L<Type::Tiny>):

    package MyTypes;
    use Type::Library -base, -declare => qw(MyInt);
    use Type::Utils;
    use Types::Standard qw(Int);

    declare MyInt, as Int, where => {$_ > 0};

    package MyClass;
    use Attribute::Contract -library => 'MyTypes';

    sub static_method : ContractRequires(ClassName, MyInt) {
    };

    ...

    MyClass->static_method(5);

=head2 IMPLEMENTATION

=head3 Inheritance

By default all the contracts are inherited. Just don't forget to C<use>
L<Attribute::Contract> in the derived class. But if no methods are overriden
then even C<using> this module is not needed.

=head3 Caching

During the compile time for every contract a Perl subroutine is built and
evaled. If the methods share the same contract they use the same checking code
reference. This speeds up the checking and saves some memory.

=head3 Error reporting

Errors are as specific as possible. On error you will get a meaningful message
and a stack trace.

=head2 SWITCHING OFF

You can switch off contract checking by specifying an environment variable
C<NO_ATTRIBUTE_CONTRACT>.

=head1 DEVELOPMENT

=head2 Repository

    http://github.com/vti/attribute-contract

=head1 AUTHOR

Viacheslav Tykhanovskyi, C<vti@cpan.org>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012-2013, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=cut


Changes
Revision history for perl module Attribute::Contract

{{$NEXT}}

0.05 2014-01-07T20:25:29Z

 - Just meta files update

0.04 2013-06-21T08:00:02Z

 - Make test less brittle

0.03 2013-06-12T09:38:42Z

 - Use cpanfile
 - Use Type::Tiny


t/tests/ContractTest.pm
package ContractTest;

use strict;
use warnings;

use base 'TestBase';

use Test::More;
use Test::Fatal;

use lib 't/tests/ContractTest';

use WithContract;
use InheritContract;
use InheritContractWithOverride;

sub handle_contract : Test {
    my $self = shift;

    ok(exception { WithContract->new->method(\1) });
}

sub inherit_contract : Test {
    my $self = shift;

    ok(exception { InheritContract->new->method(\1) });
}

sub inherit_contract_with_override : Test {
    my $self = shift;

    ok(exception { InheritContractWithOverride->new->method(\1) });
}

sub inherit_contract_with_override_in_eval : Test {
    my $self = shift;

    my $object = eval { InheritContractWithOverride->new };

    like(exception { $object->method(\1) }, qr/must be of type VALUE/);
}

sub do_not_allow_contract_change : Test {
    my $self = shift;

    like(exception { require InheritContractButChange; },
        qr/Changing contract of method 'method' in InheritContractButChange is not allowed/);
}

1;


t/tests/TypeValidatorTest.pm
package TypeValidatorTest;

use strict;
use warnings;

use base 'TestBase';

use Test::More;
use Test::Fatal;

use Attribute::Contract::TypeValidator;

sub errors : Test(3) {
    my $self = shift;

    my $tests = [
        'AN'       => [1] => qr/Unknown type 'AN'/,
        '@ANY,ANY' => [1] => qr/Array can be only the last element/,
        '%ANY,ANY' => [1] => qr/Hash can be only the last element/,
    ];

    for (my $i = 0; $i < @$tests; $i += 3) {
        my $key   = $tests->[$i];
        my $value = $tests->[$i + 1];
        my $error = $tests->[$i + 2];

        my $e = exception { $self->_build_code_ref($key)->(undef, @{$value}) };
        like($e, $error, $key);
    }
}

sub number_of_params : Test(6) {
    my $self = shift;

    my $tests = [
        'ANY'     => [1],
        'ANY,ANY' => [1, 2],
        '@ANY'    => [1],
        '@ANY'    => [1, 2],
        '@ANY'    => [1, 2, 3],
        '%ANY' => [1, 2],
    ];

    $self->_run_tests($tests);
}

sub wrong_number_of_params : Test(5) {
    my $self = shift;

    my $tests = [
        'ANY'     => [],
        'ANY'     => [1, 2],
        'ANY,ANY' => [1],
        'ANY,ANY' => [1, 2, 3],
        '%ANY'    => [1],
    ];

    $self->_run_failed_tests($tests);
}

sub values : Test(8) {
    my $self = shift;

    my $tests = [
        'VALUE'        => [1],
        'VALUE(Int)'   => [1],
        'VALUE(Int)'   => [-1],
        'VALUE(Int)'   => [+1],
        'VALUE(Str)'   => ['hi'],
        'VALUE(Float)' => [1.2],
        'VALUE(Float)' => [1.2e1],
        'VALUE(/^\d+$/)' => [0123],
    ];

    $self->_run_tests($tests);
}

sub wrong_values : Test(4) {
    my $self = shift;

    my $tests = [
        'VALUE'        => [\1],
        'VALUE(Int)'   => [1.2],
        'VALUE(Float)' => ['what?'],
        'VALUE(/^\d+$/)' => ['012a3'],
    ];

    $self->_run_failed_tests($tests);
}

sub references : Test(7) {
    my $self = shift;

    my $tests = [
        'REF'         => [\1],
        'REF'         => [{}],
        'REF(SCALAR)' => [\1],
        'REF(ARRAY)'  => [[]],
        'REF(HASH)'   => [{}],
        'REF(CODE)'   => [sub { }],
        'REF(Regexp)' => [qr/123/],
    ];

    $self->_run_tests($tests);
}

sub wrong_references : Test(9) {
    my $self = shift;

    my $tests = [
        'REF'         => [undef],
        'REF'         => [1],
        'REF(SCALAR)' => [{}],
        'REF(ARRAY)'  => [\1],
        'REF(HASH)'   => [[]],
        'REF(CODE)'   => [{}],
        'REF(Regexp)' => [sub { }],
        'REF(HASH)'   => [TypeValidatorTest->new],
        'REF'         => [TypeValidatorTest->new],
    ];

    $self->_run_failed_tests($tests);
}

sub objects : Test(2) {
    my $self = shift;

    my $tests = [
        'OBJECT'           => [__PACKAGE__->new],
        'OBJECT(TestBase)' => [__PACKAGE__->new],
    ];

    $self->_run_tests($tests);
}

sub wrong_objects : Test(3) {
    my $self = shift;

    {

        package Foo;

        sub new {
            my $class = shift;
            bless {}, $class;
        }
    }

    my $tests = [
        'OBJECT'           => [\1],
        'OBJECT'           => [qr/123/],
        'OBJECT(TestBase)' => [Foo->new],
    ];

    $self->_run_failed_tests($tests);
}

sub arrays : Test(4) {
    my $self = shift;

    my $tests = [
        '@ANY'         => [1, 2, 3],
        '@VALUE'       => [1, 2, 3],
        '@REF(SCALAR)' => [\1],
        '@REF(ARRAY)' => [[], []],
    ];

    $self->_run_tests($tests);
}

sub wrong_arrays : Test(3) {
    my $self = shift;

    my $tests = [
        '@VALUE'       => [1,  \2, 3],
        '@REF(SCALAR)' => [\1, {}],
        '@REF(ARRAY)'  => [[], 1],
    ];

    $self->_run_failed_tests($tests);
}

sub hashes : Test(4) {
    my $self = shift;

    my $tests = [
        '%ANY'         => [foo => 'bar'],
        '%VALUE'       => [foo => 'bar'],
        '%REF(SCALAR)' => [foo => \1],
        '%REF(ARRAY)'  => [foo => []],
    ];

    $self->_run_tests($tests);
}

sub wrong_hashes : Test(3) {
    my $self = shift;

    my $tests = [
        '%VALUE'       => [foo => \1],
        '%REF(SCALAR)' => [foo => 1],
        '%REF(ARRAY)'  => [foo => sub { }],
    ];

    $self->_run_failed_tests($tests);
}

sub multiple_arguments : Test(3) {
    my $self = shift;

    my $tests = [
        'ANY,VALUE,REF(SCALAR)' => [{}, 'hi', \1],
        'ANY,@ANY'              => [{}, 1,    2, 3],
        'ANY,%ANY' => [{}, foo => 'bar'],
    ];

    $self->_run_tests($tests);
}

sub optional_arguments : Test(6) {
    my $self = shift;

    my $tests = [
        'ANY,VALUE?'           => [{}, 'hi'],
        'VALUE?,VALUE?,VALUE?' => [1],
        'VALUE?,VALUE?,VALUE?' => [1,  2],
        'VALUE?,VALUE?,VALUE?' => [1, 2, 3],
        'VALUE,@ANY?'          => [1],
        'VALUE,%ANY?'          => [1],
    ];

    $self->_run_tests($tests);
}

sub alternatives : Test(7) {
    my $self = shift;

    my $tests = [
        'VALUE|REF'         => ['hi'],
        'VALUE|REF'         => [\'hi'],
        'REF(SCALAR|ARRAY)' => [\'hi'],
        'REF(SCALAR|ARRAY)' => [[]],
        '@(VALUE|REF)'      => [[], 1, \1],
        '@(REF(SCALAR)|REF(ARRAY))'      => [[], \1],
        '@(REF(SCALAR|ARRAY)|REF(HASH))' => [[], \1, {}],
    ];

    $self->_run_tests($tests);
}

sub undefs : Test(7) {
    my $self = shift;

    my $tests = [
        'ANY*'         => [undef],
        'VALUE*'       => [undef],
        'REF*'         => [undef],
        'REF*(SCALAR)' => [undef],
        'OBJECT*'      => [undef],
        '@VALUE*'      => [undef, undef],
        '%VALUE*'      => [foo => undef],
    ];

    $self->_run_tests($tests);
}

sub wrong_undefs : Test(7) {
    my $self = shift;

    my $tests = [
        'ANY'         => [undef],
        'VALUE'       => [undef],
        'REF'         => [undef],
        'REF(SCALAR)' => [undef],
        'OBJECT'      => [undef],
        '@VALUE'      => [undef, undef],
        '%VALUE'      => [foo => undef],
    ];

    $self->_run_failed_tests($tests);
}

sub _run_tests {
    my $self = shift;
    my ($tests) = @_;

    for (my $i = 0; $i < @$tests; $i += 2) {
        my $key   = $tests->[$i];
        my $value = $tests->[$i + 1];

        my $e = exception { $self->_build_code_ref($key)->(undef, @{$value}) };
        ok(!$e, $key) or diag($e);
    }
}

sub _run_failed_tests {
    my $self = shift;
    my ($tests) = @_;

    for (my $i = 0; $i < @$tests; $i += 2) {
        my $key   = $tests->[$i];
        my $value = $tests->[$i + 1];

        my $e = exception { $self->_build_code_ref($key)->(undef, @{$value}) };
        ok($e, $key);
    }
}

sub _build_code_ref {
    my $self = shift;
    my ($arguments) = @_;

    return build($arguments);
}

1;


t/tests/ContractParamsTest.pm
package ContractParamsTest;

use strict;
use warnings;

use base 'TestBase';

use Test::More;
use Test::Fatal;

use Attribute::Contract::Params;

sub throw_no_params_allowed : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref();

    like(exception { $code_ref->(undef, 1) }, qr/No params allowed/);
}

sub not_throw_when_optional_HASH : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('%ANY?');

    ok($code_ref->(undef));
}

sub not_throw_when_optional_HASH_complex : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, %ANY?');

    ok($code_ref->(undef, 1));
}

sub not_throw_when_optional_ARRAY : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('@ANY?');

    ok($code_ref->(undef));
}

sub throw_on_too_many : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR');

    like(exception { $code_ref->(undef, 1, 2) }, qr/\Q2 param(s) passed, max 1 param(s) allowed\E/);
}

sub throw_on_not_enough : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR');

    like(exception { $code_ref->(undef) }, qr/\Q0 param(s) passed, at least 1 param(s) required\E/);
}

sub not_throw_on_optional : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR,SCALAR?');

    ok(!exception { $code_ref->(undef, 1) });
}

sub not_throw_on_several_optional : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR,SCALAR?,SCALAR?');

    ok($code_ref->(undef, 1, 2));
}

sub throw_on_wrong_type : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR');

    like(exception { $code_ref->(undef, \1) }, qr/Argument 0 must be of type SCALAR/);
}

sub handle_ANY : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('ANY');

    ok($code_ref->(undef, \1));
}

sub handle_at_least_required : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, @ANY');

    like(exception {$code_ref->(undef, 1)}, qr/\Q1 param(s) passed, at least 2 param(s) required\E/);
}

sub handle_ARRAY : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, @ANY');

    ok($code_ref->(undef, 1, 'foo', 'bar'));
}

sub handle_ARRAY_with_subtype : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, @SCALAR');

    ok($code_ref->(undef, 1, 'foo', 'bar'));
}

sub throw_on_wrong_subtype : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, @SCALAR');

    like(exception {$code_ref->(undef, 1, 'foo', 'bar', \1)}, qr/Array argument 3 must be of type SCALAR/);
}

sub handle_HASH : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, %ANY');

    ok($code_ref->(undef, 1, foo => 'bar'));
}

sub throw_when_odd_parameters : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, %ANY');

    like(exception {$code_ref->(undef, 1, 'bar')}, qr/\Q2 param(s) passed, at least 3 param(s) required\E/);
}

sub throw_on_wrong_HASH_subtype : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('SCALAR, %SCALAR');

    like(exception {$code_ref->(undef, 1, foo => 'bar', baz => \1)}, qr/Hash key 'baz' value must be of type SCALAR/);
}

sub handle_OBJECT : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('OBJECT');

    ok($code_ref->(undef, ContractParamsTest->new));
}

sub handle_OBJECT_ISA : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('OBJECT(TestBase)');

    ok($code_ref->(undef, ContractParamsTest->new));
}

sub handle_ref : Test {
    my $self = shift;

    my $code_ref = $self->_build_code_ref('HASHREF');

    ok($code_ref->(undef, {}));
}

sub _build_code_ref {
    my $self = shift;
    my ($arguments) = @_;

    return Attribute::Contract::Params->modify('package', 'name',
        sub { 1 }, $arguments);
}

1;


Makefile.PL
#!/usr/bin/env perl

use 5.012;

use strict;
use warnings;

use ExtUtils::MakeMaker;

my ($mm) = $ExtUtils::MakeMaker::VERSION =~ /^([^_]+)/;

WriteMakefile(
    NAME         => 'Attribute::Contract',
    VERSION_FROM => 'lib/Attribute/Contract.pm',
    ABSTRACT     => 'Design by contract via attributes',
    AUTHOR       => 'Viacheslav Tykhanovskyi <vti@cpan.org>',

    ($mm < 6.3002 ? () : ('LICENSE' => 'artistic_2')),

    (
        $mm < 6.46
        ? ()
        : (
            META_MERGE => {
                requires  => {perl => '5.008007'},
                resources => {
                    license    => 'http://dev.perl.org/licenses/',
                    repository => 'http://github.com/vti/attribute-contract'
                },
                no_index => {directory => [qw/t/]}
            },
            META_ADD => {
                build_requires     => {},
                configure_requires => {}
            },
        )
    ),

    PREREQ_PM => {
        'Carp'         => '0',
        'Scalar::Util' => '0',
        'Exporter'     => '0',

        'Type::Tiny' => '0',

        'Test::More'  => '0',
        'Test::Fatal' => '0',
    },
    test => {TESTS => 't/*.t t/*/*.t t/*/*/*.t t/*/*/*/*.t'}
);


lib/Attribute/Contract/Utils.pm
package Attribute::Contract::Utils;

use strict;
use warnings;

use base 'Exporter';

our @EXPORT = qw(build_check);

sub build_check {
    my ($package, $name, $code_ref, $import, $attributes) = @_;

    $attributes = '' unless defined $attributes;

    my $prefix = 'use Type::Params compile => {confess => 1};';
    foreach my $key (keys %$import) {
        if ($key eq '-types') {
            my @types =
              ref $import->{$key} ? @{$import->{$key}} : ($import->{$key});
            $prefix .= "use Types::Standard qw(@types)";
        }
        elsif ($key eq '-library') {
            my $lib = $import->{$key};
            foreach my $library (ref $lib ? @$lib : $lib) {
                $prefix .= "use $library -types;";
            }
        }
    }

    my $ref = eval "$prefix; Type::Params::compile($attributes)"
      or die "Can't compile types: $@";
    return $ref;
}

1;


lib/Attribute/Contract/Params.pm
package Attribute::Contract::Params;

use strict;
use warnings;

require Carp;
use Scalar::Util qw(blessed);

$Carp::Internal{ (__PACKAGE__) }++;

sub modify {
    my $class = shift;
    my ($package, $name, $code_ref, $attributes) = @_;

    $attributes = '' unless defined $attributes;

    my @types = split /\s*,\s*/, $attributes;
    my $min_required = grep { !/\?$/ } @types;
    my $max_allowed = scalar @types;

    if (@types && $types[-1] =~ m/^%/) {
        $min_required++ unless $types[-1] =~ m/\?$/;
        $max_allowed = -1;
    }
    elsif (@types && $types[-1] =~ m/^@/) {
        $max_allowed = -1;
    }

    return sub {
        my $self = shift;

        if (@types == 0 && @_) {
            Carp::confess("No params allowed");
        }

        if ($min_required == 0 && !@_) {
            return $code_ref->($self);
        }

        Carp::confess(
            "@{[scalar(@_)]} param(s) passed, at least $min_required param(s) required"
        ) if @_ < $min_required;

        if ($max_allowed != -1) {
            Carp::confess(
                "@{[scalar(@_)]} param(s) passed, max $max_allowed param(s) allowed"
            ) if @_ > $max_allowed;
        }

        my $error = 0;
        my $pos = 0;
        foreach my $value (@_) {
            my $type = $types[$pos];

            if ($type =~ m/^(?:\@|\%)/) {
                $pos++;
                last;
            }

            my $check_code_ref = _check_code_ref($type);

            if (!$check_code_ref->($value)) {
                $error++;
                last;
            }
        }
        continue {
            $pos++;
        }

        if ($error) {
            Carp::confess("Argument $pos must be of type $types[$pos]");
        }

        if ($pos < @_ && $types[-1] =~ m/^\@(.*)/) {
            my $subtype = $1;

            my $check_code_ref = _check_code_ref($subtype);

            my $subpos = $pos - 1;
            foreach my $value (@_[$subpos .. $#_]) {
                if (!$check_code_ref->($value)) {
                    Carp::confess(
                        "Array argument $subpos must be of type $subtype");
                }

                $subpos++;
            }
        }
        elsif ($pos < @_ && $types[-1] =~ m/^\%(.*)/) {
            my $subtype = $1;

            my $check_code_ref = _check_code_ref($subtype);

            Carp::confess(
                "$package->$name: invoked with odd number of parameters")
              if ($#_ - $pos) % 2 != 0;

            my %hash = @_[($pos - 1) .. $#_];

            foreach my $key (keys %hash) {
                my $value = $hash{$key};

                if (!$check_code_ref->($value)) {
                    Carp::confess(
                        "Hash key '$key' value must be of type $subtype");
                }
            }
        }


        return $code_ref->($self, @_);
    };
}

sub _check_code_ref {
    my ($type) = @_;

    $type =~ s/\?$//;

    if ($type =~ m/^(.*?)REF$/) {
        my $ref_type = $1;
        sub { $ref_type eq ref $_[0] }
    }
    elsif ($type =~ m/^OBJECT(?:\((.*?)\))?$/) {
        my $isa = $1;
        $isa ? sub { blessed $_[0] && $_[0]->isa($isa) } : sub { blessed $_[0] }
    }
    elsif ($type eq 'SCALAR') {
        sub { !ref $_[0] }
    }
    elsif ($type eq 'ANY') {
        sub { 1 }
    }
    else {
        die "Unknown type $type";
    }
}

1;


lib/Attribute/Contract.pm
package Attribute::Contract;

use strict;
use warnings;

use 5.012;
use attributes;

our $VERSION = '0.04';

use constant NO_ATTRIBUTE_CONTRACT => $ENV{NO_ATTRIBUTE_CONTRACT};

use Scalar::Util qw(refaddr);

use Attribute::Contract::Modifier::Requires;
use Attribute::Contract::Modifier::Ensures;

BEGIN {
    use Exporter ();
    our (@ISA, @EXPORT);

    @ISA    = qw(Exporter);
    @EXPORT = qw(&MODIFY_CODE_ATTRIBUTES &FETCH_CODE_ATTRIBUTES);
}

our $CONTRACT_REQUIRES_ATTR_ALIAS = 'ContractRequires';
our $CONTRACT_ENSURES_ATTR_ALIAS  = 'ContractEnsures';

my %attrs;
my %modifiers;
my %symcache;
my %todo;
my %import;

sub contract_requires_name {
    $import{-names}->{requires} || $CONTRACT_REQUIRES_ATTR_ALIAS;
}

sub contract_ensures_name {
    $import{-names}->{ensures} || $CONTRACT_ENSURES_ATTR_ALIAS;
}

sub contract_attr_re {
    my $requires_name = contract_requires_name();
    my $ensures_name = contract_ensures_name();

    return qr/
        ^
        ($requires_name|$ensures_name)
        (?:\((.*?)\))?
        $
    /x;
}

sub import {
    return if NO_ATTRIBUTE_CONTRACT;

    my ($package) = caller;
    $todo{$package}++;

    shift;
    %import = @_;

    __PACKAGE__->export_to_level(1);
}

sub CHECK {
    return if NO_ATTRIBUTE_CONTRACT;

    foreach my $package (keys %todo) {
        foreach my $key (keys %modifiers) {
            my ($class, $method) = split /::/, $key;
            next unless $package->isa($class);

            next unless my $code_ref = $package->can($method);

            my $attrs = $modifiers{$key};

            foreach my $attr (@$attrs) {
                next unless $attr =~ contract_attr_re();

                attributes::->import($package, $code_ref, $attr);
            }
        }
    }
}

sub FETCH_CODE_ATTRIBUTES {
    my ($package, $subref) = @_;

    my $attrs = $attrs{refaddr $subref };

    return @$attrs;
}

sub MODIFY_CODE_ATTRIBUTES {
    my ($package, $code_ref, @attr) = @_;

    my $sym = findsym($package, $code_ref);
    my $name = *{$sym}{NAME};

    return if exists $attrs{refaddr $code_ref };
    return if exists $modifiers{"$package\::$name"};

    $attrs{refaddr $code_ref } = \@attr;
    $modifiers{"$package\::$name"} = \@attr;

    if (@attr) {
        no strict;
        my @isa = @{"$package\::ISA"};
        use strict;
        foreach my $isa (@isa) {
            my $key = "$isa\::$name";
            if (exists $modifiers{$key}) {

                my $base_contract = $modifiers{$key};
                my $contract = $modifiers{"$package\::$name"};

                if (@$base_contract == @$contract) {
                    next
                      if join(',', sort @$base_contract) eq
                          join(',', sort @$contract);
                }

                Carp::croak(qq{Changing contract of method '$name'}
                      . qq{ in $package is not allowed});
            }
        }
    }

    no warnings 'redefine';
    foreach my $attr (@attr) {
        next unless $attr =~ contract_attr_re();

        my $type      = $1;
        my $arguments = $2;

        my $modifier = $type eq contract_requires_name() ? 'Requires' : 'Ensures';

        my $class = __PACKAGE__ . '::Modifier::' . $modifier;

        *{$sym} = $class->modify($package, $name, $code_ref, \%import, $arguments);
    }

    return ();
}

# From Attribute::Handlers
sub findsym {
    my ($package, $ref) = @_;

    return $symcache{$package, $ref} if $symcache{$package, $ref};

    my $type = ref($ref);

    no strict 'refs';
    foreach my $sym (values %{$package . "::"}) {
        use strict;
        next unless ref(\$sym) eq 'GLOB';

        return $symcache{$package, $ref} = \$sym
          if *{$sym}{$type} && *{$sym}{$type} == $ref;
    }

    return;
}

1;
__END__

=head1 NAME

Attribute::Contract - Design by contract via Perl attributes

=head1 SYNOPSIS

    package Interface;
    use Attribute::Contract -types => [qw/Str slurpy ArrayRef/];

    sub do_smth :ContractRequires(Str, slurpy ArrayRef[Str]) :ContractEnsures(Str) {
        my $self = shift;
        my ($input_string, $array_ref_of_strings) = @_;

        return '...';
    }

    package Implementation;
    use base 'Interface';
    use Attribute::Contract;

    sub do_smth {
        my $self = shift;
        my ($input_string, $array_ref_of_strings) = @_;

        return 'ok';
    }

    Implementation->do_smth('hi', 'there'); # works

    Implementation->do_smth();              # croaks!
    Implementation->do_smth(sub {});        # croaks!

=head1 DESCRIPTION

L<Attribute::Contract> by using Perl attributes allows you to specify contract
(L<Design by Contract|http://en.wikipedia.org/wiki/Design_by_contract>) for
every method in your class. You can check incoming and outgoing values by
specifying C<ContractRequires> and C<ContractEnsures> attributes.

It's the most useful for interfaces or abstract classes when you want to control
whether your implementation follows the same interface and respects the Liskov
substitution principle.

This module uses L<Type::Tiny> underneath so all the checks is done via that
module. Check it out for more documention on type validation.

Why attributes? They feel and look natural and are applied during compile time.

=head2 IMPORTING

When using L<Attribute::Contract> one may want to import various types in order
to check them in attributes. Types themselves are not imported into the current
module but rather used when compiling attributes.

=head3 Types

    package MyClass;
    use Attribute::Contract -types => [qw/ClassName Str/];

    sub static_method : ContractRequires(ClassName, Str) {
    };

    ...

    MyClass->static_method('string');

=head3 Type libraries

When types are complex or the description is too long attributes might get not
very readable. In this case one can use type libraries (implemented again via
L<Type::Tiny>):

    package MyTypes;
    use Type::Library -base, -declare => qw(MyInt);
    use Type::Utils;
    use Types::Standard qw(Int);

    declare MyInt, as Int, where => {$_ > 0};

    package MyClass;
    use Attribute::Contract -library => 'MyTypes';

    sub static_method : ContractRequires(ClassName, MyInt) {
    }

    ...

    MyClass->static_method(5);

=head3 Aliasing

If you don't like C<ContractRequires> and C<ContractEnsures> you can set your
own names:

    use Attribute::Contract -names => {requires => 'In', ensures => 'Out'}

    sub method : In(ClassName, Str) Out(Str) {
    }

=head2 IMPLEMENTATION

=head3 Inheritance

By default all the contracts are inherited. Just don't forget to C<use>
L<Attribute::Contract> in the derived class. But if no methods are overriden
then even C<using> this module is not needed.

=head3 Caching

During the compile time for every contract a Perl subroutine is built and
evaled. If the methods share the same contract they use the same checking code
reference. This speeds up the checking and saves some memory.

=head3 Error reporting

Errors are as specific as possible. On error you will get a meaningful message
and a stack trace.

=head2 SWITCHING OFF

You can switch off contract checking by specifying an environment variable
C<NO_ATTRIBUTE_CONTRACT>.

=head1 DEVELOPMENT

=head2 Repository

    http://github.com/vti/attribute-contract

=head1 AUTHOR

Viacheslav Tykhanovskyi, C<vti@cpan.org>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012-2013, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=cut
